### ISSUE TITLE
FAILED ert_tests/ensemble_evaluator/test_ensemble_legacy.py::test_run_and_cancel_legacy_ensemble

### ISSUE THREAD

**Issue by frode-aarstad** [equinor/ert#2844]
**Describe the bug**
Failed github action on PR.

https://github.com/equinor/ert/actions/runs/1793572277

`FAILED ert_tests/ensemble_evaluator/test_ensemble_legacy.py::test_run_and_cancel_legacy_ensemble`
```

 =================================== FAILURES ===================================
_____________________ test_run_and_cancel_legacy_ensemble ______________________

tmpdir = local('/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pytest-of-runner/pytest-0/test_run_and_cancel_legacy_ens0')
make_ensemble_builder = <function make_ensemble_builder.<locals>._make_ensemble_builder at 0x133ab0a60>

    @pytest.mark.timeout(60)
    def test_run_and_cancel_legacy_ensemble(tmpdir, make_ensemble_builder):
        num_reals = 10
        custom_port_range = range(1024, 65535)
        with tmpdir.as_cwd():
            ensemble = make_ensemble_builder(tmpdir, num_reals, 2, job_sleep=5).build()
            config = EvaluatorServerConfig(custom_port_range=custom_port_range)
    
            evaluator = EnsembleEvaluator(ensemble, config, 0, ee_id="1")
    
            with evaluator.run() as mon:
                cancel = True
                for _ in mon.track():
                    if cancel:
                        mon.signal_cancel()
                        cancel = False
    
>           assert evaluator._ensemble.get_status() == state.ENSEMBLE_STATE_CANCELLED
E           AssertionError: assert 'Stopped' == 'Cancelled'
E             - Cancelled
E             + Stopped

ert_tests/ensemble_evaluator/test_ensemble_legacy.py:52: AssertionError
```

Also logged 

    WARNING  ert_shared.ensemble_evaluator.ensemble.base:base.py:53 Illegal state transition from Cancelled to Stopped

Which is a hint that perhaps this test was always flaky, or that cancelling doesn't work properly.

**To Reproduce**
1. Push to a PR

**Expected behavior**
Dont fail

**Enviromment**
 - 3.8 macos-latest

**Comment by BjarneHerland** [comment]
This happens when a STOP-event is dispatched after a CANCEL-event. Adding the right barriers and delays can reproduce this issue consistently.

It happens when execution gets to this block
https://github.com/equinor/ert/blob/2c0c38bdd4ed5b6e0181c66baae930e93eb81ce6/ert_shared/ensemble_evaluator/ensemble/legacy.py#L158-L170
but this block
https://github.com/equinor/ert/blob/2c0c38bdd4ed5b6e0181c66baae930e93eb81ce6/ert_shared/ensemble_evaluator/ensemble/legacy.py#L208-L222
has been executed first.

This might be caused by the coroutine wrapped by `self._aggregate_future`
https://github.com/equinor/ert/blob/2c0c38bdd4ed5b6e0181c66baae930e93eb81ce6/ert_shared/ensemble_evaluator/ensemble/legacy.py#L141-L151
catching and ignoring the `CancelledError` which is delivered when calling `self._aggregate_future.cancel()`, causing it to exit normally even if cancel was called.

**Comment by BjarneHerland** [comment]
This issue can in fact be reproduced with pytest-repeat, failing 1-2 times out of 100 for me locally. The root-problem is in line 458 in the job-queue here
https://github.com/equinor/ert/blob/82575f6535e22d91e343def18844f9bfab746fc7/res/job_queue/queue.py#L445-L475
where the queue decides to quit based on the flag `self._stopped`, dropping to line 469 and **not re-raising** the `CancelledError`.

`JobQueue._stopped` is set by `JobQueue.kill_all_jobs()` called from `_LegagyEnsemble.cancel()` (line 200) prior to cancelling `self. _aggregate_future` (line 205) here
https://github.com/equinor/ert/blob/82575f6535e22d91e343def18844f9bfab746fc7/ert_shared/ensemble_evaluator/ensemble/legacy.py#L195-L216

Hence, the problem occurs when line 458 executes after line 200 but before line 205 (or at least before `CancelledError` is delivered to the event-loop). This makes `_LegacyEnsemble` dispatch a STOP-event since the jobs appear to have finished normally.

Because `_LegagyEnsemble.cancel()` runs asynchrounously and dispatches a CANCEL-event (line 216), we risk a sequence of events where CANCEL comes first, then STOP, which triggers the warning quoted above, and makes the final state of the ensemble STOPPED causing the test to fail.

By not calling `JobQueue.kill_all_jobs()` from `_LegagyEnsemble.cancel()` (comment out line 200) and making line 465 unconditional, this test pass 100 of 100 repetitions.

**Comment by BjarneHerland** [comment]
Hmm...  when repeating the experiment with 100 repetitions, I see that the enhanced verifications introduced in https://github.com/equinor/ert/commit/82575f6535e22d91e343def18844f9bfab746fc7 sometimes fail also with patches indicated above.

The sometimes-failing asserts are these
https://github.com/equinor/ert/blob/82575f6535e22d91e343def18844f9bfab746fc7/tests/ert_tests/ensemble_evaluator/test_ensemble_legacy.py#L61-L63

@jondequinor Is the legacy-ensemble able to kill jobs which have started running? I.e. do these extra verifications make sense in this cancel-test?

**Comment by jondequinor** [comment]
Yeah, but if you ask to kill something, it's not happening immediately, but in 1-seconds intervals:

`_job_monitor` has a `while-sleep(1)` loop and if `_should_be_killed` is true (as in the case of `kill_all_jobs`), `kill()` is called on the driver, which for local runs (i.e. tests) is a `SIGTERM`.

So it makes sense if all jobs are overly slow to produce this output. If they are producing output +- 1 seconds after you press cancel—who knows? It's a race.

**Comment by jondequinor** [comment]
I'm not able to reproduce this locally, but how would this diff fare?

```diff
diff --git a/res/job_queue/queue.py b/res/job_queue/queue.py
index e1b416369..9d77b9ac3 100644
--- a/res/job_queue/queue.py
+++ b/res/job_queue/queue.py
@@ -444,10 +444,10 @@ class JobQueue(BaseCClass):

             try:
                 while True:
-                    self.launch_jobs(pool_sema)
-
                     await asyncio.sleep(1)

+                    self.launch_jobs(pool_sema)
+
                     if evaluators is not None:
                         for func in evaluators:
                             func()
@@ -455,8 +455,6 @@ class JobQueue(BaseCClass):
                     await JobQueue._publish_changes(
                         ee_id, self.changes_after_transition(), websocket
                     )
-                    if not self.is_active() or self.stopped:
-                        break
             except asyncio.CancelledError:
                 if self.stopped:
                     logger.debug(
```

Ignore that jobs are stopped and rely on `await asyncio.sleep` raising cancelled? Might be worth a shot.

**Comment by jondequinor** [comment]
But I think both the evaluator and the legacy ensemble is wrong here:

- the evaluator should be fine with `CANCELLED` then `STOPPED`. It should just ignore `STOPPED`. Display a warning and move on.
- the legacy ensemble is absolute crap (and I wrote a lot of it). Its API is sync, it does a lot of async stuff, but the majority of the queue stuff is sync. Guido would actually sue us if he knew.

So a fix is https://github.com/equinor/ert/issues/3124 and ignoring the funky state transition.

**Comment by BjarneHerland** [comment]
> But I think both the evaluator and the legacy ensemble is wrong here:
> 
> * the evaluator should be fine with `CANCELLED` then `STOPPED`. It should just ignore `STOPPED`. Display a warning and move on.
> * the legacy ensemble is absolute crap (and I wrote a lot of it). Its API is sync, it does a lot of async stuff, but the majority of the queue stuff is sync. Guido would actually sue us if he knew.
> 
> So a fix is #3124 and ignoring the funky state transition. WDYT?

I don't disagree with any of the above :grin: but have suggested a lightweight fix for our immediate flaky-problems in #3125.
